### PR TITLE
Cleanup RAND_load_file,RAND_write_file

### PR DESCRIPTION
**Pull request by richsalz**
Document an internal assumption that these are only for use with files,
and return an error if not. That made the code much simpler.
Also only write 256 bytes, we don't need more than that from a security
perspective.

This was pulled out from #3862.

### CODE REVIEW

File: crypto/rand/randfile.c [hunk 1]
```diff
@@ -106,112 +81,71 @@ static __FILE_ptr32 (*const vms_fopen)(const char *, const char *, ...) =
...
-    if (S_ISBLK(sb.st_mode) || S_ISCHR(sb.st_mode)) {
-        /*
-         * this file is a device. we don't want read an infinite number of
-         * bytes from a random device, nor do we want to use buffered I/O
-         * because we will waste system entropy.
-         */
-        bytes = (bytes == -1) ? 2048 : bytes; /* ok, is 2048 enough? */
-        setbuf(in, NULL); /* don't do buffered reads */
+    if (stat(file, &sb) < 0 || !S_ISREG(sb.st_mode)) {
+        RANDerr(RAND_F_RAND_LOAD_FILE, RAND_R_NOT_IFREG);
```
Comment (levitte) [inline crypto/rand/randfile.c]: Please make that `RAND_R_NOT_A_REGULAR_FILE`.  You might also want to add something like this:

```
        ERR_add_error_data(2, "Filename=", file);
```

File: crypto/rand/randfile.c [hunk 2]
```diff
@@ -106,112 +81,71 @@ static __FILE_ptr32 (*const vms_fopen)(const char *, const char *, ...) =
...
+        return -1;
     }
-# endif
 #endif
-    for (;;) {
+    if ((in = openssl_fopen(file, "rb")) == NULL) {
+        RANDerr(RAND_F_RAND_LOAD_FILE, RAND_R_CANNOT_FOPEN);
```
Comment (levitte) [inline crypto/rand/randfile.c]: `RAND_R_CANNOT_OPEN_FILE`, and perhaps same additional data as above?

File: crypto/rand/randfile.c [hunk 3]
```diff
@@ -106,112 +81,71 @@ static __FILE_ptr32 (*const vms_fopen)(const char *, const char *, ...) =
...
-            /*
-             * this file is a device. we don't write back to it. we
-             * "succeed" on the assumption this is some sort of random
-             * device. Otherwise attempting to write to and chmod the device
-             * causes problems.
-             */
-            return 1;
-        }
+    if (stat(file, &sb) >= 0 && !S_ISREG(sb.st_mode)) {
+        RANDerr(RAND_F_RAND_WRITE_FILE, RAND_R_NOT_IFREG);
```
Comment (levitte) [inline crypto/rand/randfile.c]: Same thing here

**Comment by richsalz** [comment]
Okay, additional commit pushed that does your suggestions, @levitte

**Comment by levitte** [comment]
Cool. I'll do a double check on VMS tomorrow (way too tired to do it now)

**Comment by richsalz** [comment]
so, @levitte did I break VMS or did it work? :)

**Comment by levitte** [comment]
[ahem] I was distracted...  trying it now

**Comment by richsalz** [comment]
You’re not allowed to have a life ☺

**Comment by levitte** [comment]
Stop scaring me ;-)

**Comment by richsalz** [comment]
Thanks!  commit 9ee344f

File: crypto/rand/randfile.c [hunk 4]
```diff
@@ -72,32 +56,22 @@
...
 #  pragma pointer_size restore
-/*
- * On VMS, setbuf() will only take 32-bit pointers, and a compilation
- * with /POINTER_SIZE=64 will give off a MAYLOSEDATA2 warning here.
- * Since we know that the FILE* really is a 32-bit pointer expanded to
- * 64 bits, we also know it's safe to convert it back to a 32-bit pointer.
- * As for the buffer parameter, we only use NULL here, so that passes as
- * well...
- */
-#  define setbuf(fp,buf) (setbuf)((__FILE_ptr32)(fp), (char_ptr32)(buf))
```
Comment (dot-asm) [inline crypto/rand/randfile.c]: If this is removed, then one doesn't need `typedef char *char_ptr32`...

**Comment by richsalz** [comment]
Good point.  New PR coming up.

**Pull request closed without merging**